### PR TITLE
Add abstention metrics to retrieval and reranking tasks

### PR DESCRIPTION
**Pull request by hgissbkh**
Hi @KennethEnevoldsen,
As discussed in PR [https://github.com/embeddings-benchmark/mteb/pull/841](https://github.com/embeddings-benchmark/mteb/pull/841), the most up-to-date version of our code is now here!

### CODE REVIEW

**Comment by hgissbkh** [comment]
As promised, here's a small evaluation table of three reranking tasks for the intfloat/multilingual-e5-small model. As a reminder, the closer the nAUC metrics are to 1, the closer the abstention curve is to the oracle.

|                |   AskUbuntuDupQuestions |   MIRACLReranking_deu-Latn |   MIRACLReranking_spa-Latn |
|:---------------|------------------------:|---------------------------:|---------------------------:|
| map            |               0.568003  |                   0.636789 |                  0.734434  |
| mrr            |               0.709497  |                   0.725021 |                  0.818764  |
| nAUC_map_max   |               0.266341  |                   0.244547 |                  0.17054   |
| nAUC_map_std   |               0.202235  |                   0.260974 |                  0.138472  |
| nAUC_map_diff1 |               0.0492681 |                   0.169382 |                  0.0430034 |
| nAUC_mrr_max   |               0.37968   |                   0.262468 |                  0.241771  |
| nAUC_mrr_std   |               0.240068  |                   0.302836 |                  0.140447  |
| nAUC_mrr_diff1 |               0.121867  |                   0.192988 |                  0.0984651 |

File: mteb/evaluation/evaluators/utils.py [hunk 1]
```diff
@@ -264,3 +292,107 @@ def download(url: str, fname: str):
         for data in resp.iter_content(chunk_size=1024):
             size = file.write(data)
             bar.update(size)
+
+
+def confidence_scores(sim_scores: List[float]) -> Dict[str, float]:
+    """Computes confidence scores for a single instance = (query, positives, negatives)
+
+    Args:
+        sim_scores (`List[float]` of shape `(num_pos+num_neg,)`): Query-documents similarity scores
```
Comment (KennethEnevoldsen) [inline mteb/evaluation/evaluators/utils.py:301]: remove (...) - move the shape information into description of arg

File: mteb/evaluation/evaluators/RerankingEvaluator.py [hunk 2]
```diff
@@ -202,30 +221,89 @@ def _encode_unique_texts(self, all_texts, encode_queries_func):
...
+
+        return sim_scores
+
+    def _compute_metrics_instance(
+        self, sim_scores: torch.Tensor, is_relevant: List[bool]
+    ) -> Dict[str, float]:
+        """Computes metrics for a single instance = (query, positives, negatives)
+
+        Args:
+            sim_scores (`torch.Tensor of shape `(num_pos+num_neg,)`): Query-documents similarity scores
```
Comment (KennethEnevoldsen) [inline mteb/evaluation/evaluators/RerankingEvaluator.py:249]:
```suggestion
            sim_scores: Matrix of Query-documents similarity scores of  shape `(num_pos+num_neg,)
```
general correction

**Review by KennethEnevoldsen — APPROVED** [review]
Thanks for this only a few minor things but otherwise we can merge

Edit: Once merged can I ask you to provide a description of the abstention metrics in the section on new task types (appendix - it might not be there yet but then feel free to create it) - see #595 for link

File: mteb/evaluation/evaluators/RerankingEvaluator.py (hunk 2 above)
Comment (hgissbkh) [reply mteb/evaluation/evaluators/RerankingEvaluator.py]: Should be fine now, just pushed the updated docstrings!

**Comment by KennethEnevoldsen** [comment]
Auto-merge enabled. Thanks for the PR(s) and the fruitful debate along with it as well.

File: docs/mmteb/points/854.jsonl [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{"GitHub": "ManuelFay", "Bug fixes": 13,  "New task": 5}
+{"GitHub": "hgissbkh", "Bug fixes": 13, "New task": 5}
+{"GitHub": "KennethEnevoldsen", "PR Review": 2}
+{"GitHub": "imenelydiaker", "PR Review": 2}
+{"GitHub": "orionw", "PR Review": 2}
```
Comment (KennethEnevoldsen) [inline docs/mmteb/points/854.jsonl:5]:
```suggestion
{"GitHub": "imenelydiaker", "Review PR": 2}
{"GitHub": "orionw", "Review PR": 2}
```

**Review by KennethEnevoldsen — APPROVED** [review]

**Comment by hgissbkh** [comment]
Hey @KennethEnevoldsen! 
Just added a description of the abstention metric in the overleaf (appendix E - New Tasks).
Lmk if the tone is not right, I have tried not to put too many mathematical notations to keep it light.
And thanks again for the merge!

**Pull request merged**